### ISSUE TITLE
BUG: scipy's pdist(X, metric='dice') VS scipy.spatial.distance.dice() produce different results

### ISSUE THREAD

**Issue by PhillipMaire** [scipy/scipy#17703]
### Describe your issue.

Please forgive me if this is just a lack of my understanding of the difference between the two function but I can't seem to find why these would be different. the code should make it clear, but basically I am confused why there is a different result (for the last set `[1, 0, 0], [2, 0, 0]` for these two metrics that I assume should be the same. 

thanks!

### Reproducing Code Example

```python
#pdist

from scipy.spatial.distance import pdist

X = [[1, 0, 0], [0, 1, 0]]
tmp1 = pdist(X, metric='dice')
print(tmp1)
# [1.]
X = [[1, 0, 0], [1, 1, 0]]
tmp1 = pdist(X, metric='dice')
print(tmp1)
# [0.33333333]
X = [[1, 0, 0], [2, 0, 0]]
tmp1 = pdist(X, metric='dice')
print(tmp1)
# [0.]

#dice

from scipy.spatial import distance
tmp1 = distance.dice([1, 0, 0], [0, 1, 0])
print(tmp1)
# 1.0
tmp1 = distance.dice([1, 0, 0], [1, 1, 0])
print(tmp1)
# 0.3333333333333333
tmp1 = distance.dice([1, 0, 0], [2, 0, 0])
print(tmp1)
# -0.3333333333333333
```

### Error message

```shell
None
```

### SciPy/NumPy/Python version information

1.8.0 1.21.6 sys.version_info(major=3, minor=8, micro=4, releaselevel='final', serial=0)

**Comment by tupui** [comment]
Hi @PhillipMaire thank you for reporting. This is because we have a type conversion happening for `dice`. As the doc of `dice` suggests, parameters should be booleans. In `dice` itself, it's allowed to use something else than booleans and here is the difference. `pdist` is wrapping the function and doing some input conversion.

I am not sure why we have a difference here. Looks like an oversight and I think we should either allow other types to be passed to `pdist` (and `cdist`, also here the doc wrongly start by saying inputs are converted to float) for `dice`, or not and update the doc to not show improper usages.

**Comment by PhillipMaire** [comment]
ok I see, thanks for the clarification. After realized a mistake I made I actually I don't need this metric anymore, but glad I could help identify a bug.

**Comment by tylerjereddy** [comment]
This is kind of a stats concept, right? I believe `scikit-learn` and `pytorch` both provide either `dice` or the (equivalent for boolean?) F1 score, in case there are any cues to take from them.

**Comment by tupui** [comment]
AFAICT, they say it's acting on booleans

> Metrics intended for boolean-valued vector spaces: Any nonzero entry is evaluated to “True”. 

So I guess we could do the same here and do some input validation in `dice` and cast to bool as we do for `pdist/cdist`. We could start with raising a deprecation warning if non 0, 1, bool values are present. WDYT?

**Comment by peterbell10** [comment]
The python metrics support float values between 0 and 1 as "fuzzy" bools. It would be more consistent to fall back to the slow path instead of casting to bool.

**Comment by tupui** [comment]
Thank you @peterbell10. Could you clarify? I don't understand what is your concrete suggestion.

**Comment by tupui** [comment]
mmm ok I suppose we could check that during the `_validate_cdist_input/_validate_pdist_input` step. The function would also return the metric function to use (it's already taking in the metric dict.)

Does that sound correct?

**Comment by peterbell10** [comment]
I think handling this in the wrapper is good. Type promotion is okay, and AFAIK all the `_distance_pybind` functions promote to floating point. We really only want to fall back when the input types cannot safely be cast to the supported type, which is mainly boolean metrics.

**Comment by tupui** [comment]
Ok attempting to put a good first issue there. I think it could be approachable with some guidance. Feel free to remove if I am too ambitious there 😅

**Comment by LiorHaimGit** [comment]
I would be happy to attempt to repair it.

**Comment by tsabbir96** [comment]
Hey @tupui Do you mind if I take on this issue, it seems challenging but would be a good contribution if I can fix it successfully (by some guidance of course, haha )

**Comment by tupui** [comment]
Hi @tsabbir96 I think @peterbell10 wanted to fix this along with #17538. Can you confirm Peter?

**Comment by juanis2112** [comment]
I attempted to fix this issue as part of the SciPyConf2023 sprint but I couldn't reproduce the issue with `scipy=1.11.1` or `scipy=1.11.0`. However, I was able to reproduce it with `scipy=1.10.2`. I think it was fixed in #17538 so I'm going to use the example here to write a test for it in order to close this issue.

**Comment by tylerjereddy** [comment]
@juanis2112 Agreed, that is helpful to know, closing the issue, thanks!
